cmd/generic-handlers.go

```diff
@@ -723,6 +723,20 @@ func (l rateLimit) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	l.handler.ServeHTTP(w, r)
 }
 
+type requestIDHeaderHandler struct {
+	handler http.Handler
+}
+
+func addrequestIDHeaders(h http.Handler) http.Handler {
+	return requestIDHeaderHandler{handler: h}
+}
+
+func (s requestIDHeaderHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Set unique request ID for each response.
+	w.Header().Set(responseRequestIDKey, mustGetRequestID(UTCNow()))
+	s.handler.ServeHTTP(w, r)
+}
```

Comment:
shouldn't the requestId you generate here be used by http_tracer.go while logging request and response instead of generating a reqId dynamically?  X-amz-Request-Id in the response below should be traceable on the log to be usable.
```
REQUEST (objectAPIHandlers).ListObjectsV1Handler-fm] [153204311.516598] [2018-07-19 16:31:55 -0700]
GET /test/?delimiter=%2F&max-keys=1000&prefix=weer
Host: localhost:9000
User-Agent: Minio (linux; amd64) minio-go/v6.0.6 mc/DEVELOPMENT.GOGET
Authorization: AWS4-HMAC-SHA256 Credential=minio/20180719/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=dc859059c87fb913d4a2ad213554335cc67ab04d2c5a3ef88448ef3d858433c7
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20180719T233155Z


[RESPONSE] [153204311.516598] [2018-07-19 16:31:55 -0700]
200 OK
Content-Security-Policy: block-all-mixed-content
X-Amz-Request-Id: 1542E9189C0C6CEB
Server: Minio/DEVELOPMENT.GOGET (linux; amd64)
Accept-Ranges: bytes
Content-Type: application/xml
Vary: Origin
X-Xss-Protection: 1; mode=block

<?xml version="1.0" encoding="UTF-8"?>
```

Reply:
I see it being returned as part of the response. Not quite understand, what you are saying here.

Here is a sample output . I modified the code to also print the RequesID once it is generated.

```
1542ECA94526B210
[REQUEST (objectAPIHandlers).GetBucketLocationHandler-fm] [153204703.501472] [2018-07-19 17:37:15 -0700]
GET /unwammamaa/?location=
Host: localhost:9000
User-Agent: Minio (linux; amd64) minio-go/v6.0.6 mc/2018-07-17T17:52:17Z
Authorization: AWS4-HMAC-SHA256 Credential=minio/20180720/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=84359c936e9e41a529433a67a512ecef39ddfaacbe27c4e7057d224ac21b8b07
X-Amz-Content-Sha256: UNSIGNED-PAYLOAD
X-Amz-Date: 20180720T003714Z


[RESPONSE] [153204703.501472] [2018-07-19 17:37:15 -0700]
404 Not Found
Vary: Origin
X-Xss-Protection: 1; mode=block
Content-Security-Policy: block-all-mixed-content
X-Amz-Request-Id: 1542ECA94526B210
Server: Minio/DEVELOPMENT.2018-07-20T00-36-54Z (linux; amd64)
Accept-Ranges: bytes
Content-Type: application/xml

<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>NoSuchBucket</Code><Message>The specified bucket does not exist</Message><Key></Key><BucketName></BucketName><Resource>/unwammamaa/</Resource><RequestId>1542ECA94526B210</RequestId><HostId>3L137</HostId></Error>

1542ECA945923E4F
[REQUEST (objectAPIHandlers).ListBucketsHandler-fm] [153204703.502026] [2018-07-19 17:37:15 -0700]
GET /
Host: localhost:9000
User-Agent: Minio (linux; amd64) minio-go/v6.0.6 mc/2018-07-17T17:52:17Z
Authorization: AWS4-HMAC-SHA256 Credential=minio/20180720/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=8080eddd3ef2b7e34f0dbd032fb9f2498648bf9b72927e0df424c6b23b439d88
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20180720T003715Z


[RESPONSE] [153204703.502026] [2018-07-19 17:37:15 -0700]
200 OK
Vary: Origin
X-Xss-Protection: 1; mode=block
Content-Security-Policy: block-all-mixed-content
X-Amz-Request-Id: 1542ECA945923E4F
Server: Minio/DEVELOPMENT.2018-07-20T00-36-54Z (linux; amd64)
Accept-Ranges: bytes
Content-Type: application/xml

<?xml version="1.0" encoding="UTF-8"?>
<ListAllMyBucketsResult ....
```

There are two http calls logged here. The value that is printed above the line `[REQUEST..` in the first call is `1542ECA94526B210`, if you look at the corresponding `x-amz-request-id` in the reponse, you can see that it is `X-Amz-Request-Id: 1542ECA94526B210`

You can see that the reqid is generated only once.

Reply:
The x-amz-request-id 1542ECA945923E4F should be the one printed in http trace instead of 153204703.501472, so that if a user has issues with a particular requestId, it can be debugged from log. Look at [TraceReqHandlerFunc](https://github.com/minio/minio/blob/866dffcd6229eca21c30c64cf2b85290e1852222/pkg/handlers/http-tracer.go) line # 138. Wouldn't you use the reqId already present in the http.request/http.responseWriter instead of generating a different reqID for the log?

Reply:
nvm - I misunderstood the http trace purpose here.

Reply:
Now that I see the code, I see what you mean. That is not the request ID that we are talking about here. I don't know why the code calls it `reqID`, but it is to match the `request` with the corresponding `response`.
```
[REQUEST (objectAPIHandlers).ListBucketsHandler-fm] [153204703.502026] [20..
```
```
[RESPONSE] [153204703.502026] [2018-07-19 17:37:15 -0700]
200 OK
Vary: Origin
X-Xss-Protection: 1; mode=block
Content-Security-Policy: block-all-mixed-content
X-Amz-Request-Id: 1542ECA945923E4F
```

This could be for the case where lots of requests are coming to the server and we have `HTTP TRACE` turned on. If there are interleaving `requests` and `responses`, then this can be used as an identifier.

This PR does not do anything about that. It is concerned only about `X-Amz-Request-Id`. Here you should check if `X-Amz-Request-Id: 1542ECA945923E4F` is same as the one that was printed before `REQUEST...`